### ISSUE TITLE
NuGet 3.4 fails to install package ending with zero

### ISSUE THREAD

**Issue by NN---** [NuGet/Home#2298]
I have package "A" with version 1.1.1.0 .

Prior 3.4 this command works well:

```
nuget install A -version 1.1.1.0
```

With NuGet 3.4 RC I get:

```
An error occurred while retrieving package metadata for 'A.1.1.1' from source 'N'.
  An error occurred while retrieving package metadata for 'A.1.1.1' from source 'N'.
  Data at the root level is invalid. Line 1, position 1.
```

**Comment by emgarten** [comment]
What type of source are you installing from? Online v2? Online v3? Local disk?

**Comment by NN---** [comment]
I use ProGet server (http://inedo.com/proget) which supports v2 source.

**Comment by emgarten** [comment]
@NN--- would you be able to try this using NuGet in Visual Studio with the same proget server and let me know if it works there? 

https://dist.nuget.org/visualstudio-2015-vsix/v3.4.0-rc/NuGet.Tools.vsix

There have been some changes in the V2 protocol and it would help to know if this is a problem in the install command, or if NuGet has a general problem with proget servers and non-normalized package versions.

**Comment by NN---** [comment]
Both don't work.
I suspected VS extension and check also the command line tool.

ProGet server can be downloaded and installed for free if you want to check it directly.

**Comment by zhili1208** [comment]
I didn't repro this with ProGet on my machine, so can you specify -Verbosity in command Line, then you can see all http call. try the http call, I guess it returns invalid xml.

**Comment by emgarten** [comment]
`nuget install A -version 1.1.1.0 --verbosity detailed`

**Comment by NN---** [comment]
This what I get with nuget 3.4 RC:

```
System.AggregateException: One or more errors occurred. ---> System.InvalidOperationException: An error occurred while retrieving package metadata for 'A.1.1.1' from source 'ProGet'. ---> NuGet.Protocol.Core.Types.FatalProtocolException: An error occurred while retrieving package metadata for 'A.1.1.1' from source 'ProGet'. ---> System.Xml.XmlException: Data at the root level is invalid. Line 1, position 1.
   at System.Xml.XmlTextReaderImpl.Throw(Exception e)
   at System.Xml.XmlTextReaderImpl.Throw(String res, String arg)
   at System.Xml.XmlTextReaderImpl.ParseRootLevelWhitespace()
   at System.Xml.XmlTextReaderImpl.ParseDocumentContent()
   at System.Xml.XmlTextReaderImpl.Read()
   at System.Xml.Linq.XDocument.Load(XmlReader reader, LoadOptions options)
   at NuGet.Protocol.V2FeedParser.LoadXml(Stream stream)
   at NuGet.Protocol.V2FeedParser.<QueryV2Feed>d__49.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.Protocol.V2FeedParser.<GetPackage>d__40.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.Protocol.DependencyInfoResourceV2Feed.<ResolvePackage>d__4.MoveNext()
   --- End of inner exception stack trace ---
   at NuGet.Protocol.DependencyInfoResourceV2Feed.<ResolvePackage>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.PackageManagement.ResolverGather.<GatherPackageFromSourceAsync>d__27.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at NuGet.PackageManagement.ResolverGather.<GatherPackageAsync>d__26.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.PackageManagement.ResolverGather.<>c__DisplayClass25_0.<<StartWorkerTasks>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.PackageManagement.ResolverGather.<ProcessResultsAsync>d__24.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.PackageManagement.ResolverGather.<StartTasksAndProcessWork>d__21.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at NuGet.PackageManagement.ResolverGather.<GatherAsync>d__20.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.PackageManagement.ResolverGather.<GatherAsync>d__19.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at NuGet.PackageManagement.NuGetPackageManager.<PreviewInstallPackageAsync>d__43.MoveNext()
   --- End of inner exception stack trace ---
   at NuGet.PackageManagement.NuGetPackageManager.<PreviewInstallPackageAsync>d__43.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.PackageManagement.NuGetPackageManager.<InstallPackageAsync>d__28.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at NuGet.CommandLine.InstallCommand.<InstallPackage>d__34.MoveNext()
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at NuGet.CommandLine.Command.Execute()
   at NuGet.CommandLine.Program.MainCore(String workingDirectory, String[] args)
---> (Inner Exception #0) System.InvalidOperationException: An error occurred while retrieving package metadata for 'A.1.1.1' from source 'ProGet'. ---> NuGet.Protocol.Core.Types.FatalProtocolException: An error occurred while retrieving package metadata for 'A.1.1.1' from source 'ProGet'. ---> System.Xml.XmlException: Data at the root level is invalid. Line 1, position 1.
   at System.Xml.XmlTextReaderImpl.Throw(Exception e)
   at System.Xml.XmlTextReaderImpl.Throw(String res, String arg)
   at System.Xml.XmlTextReaderImpl.ParseRootLevelWhitespace()
   at System.Xml.XmlTextReaderImpl.ParseDocumentContent()
   at System.Xml.XmlTextReaderImpl.Read()
   at System.Xml.Linq.XDocument.Load(XmlReader reader, LoadOptions options)
   at NuGet.Protocol.V2FeedParser.LoadXml(Stream stream)
   at NuGet.Protocol.V2FeedParser.<QueryV2Feed>d__49.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.Protocol.V2FeedParser.<GetPackage>d__40.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.Protocol.DependencyInfoResourceV2Feed.<ResolvePackage>d__4.MoveNext()
   --- End of inner exception stack trace ---
   at NuGet.Protocol.DependencyInfoResourceV2Feed.<ResolvePackage>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.PackageManagement.ResolverGather.<GatherPackageFromSourceAsync>d__27.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at NuGet.PackageManagement.ResolverGather.<GatherPackageAsync>d__26.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.PackageManagement.ResolverGather.<>c__DisplayClass25_0.<<StartWorkerTasks>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.PackageManagement.ResolverGather.<ProcessResultsAsync>d__24.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.PackageManagement.ResolverGather.<StartTasksAndProcessWork>d__21.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at NuGet.PackageManagement.ResolverGather.<GatherAsync>d__20.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.PackageManagement.ResolverGather.<GatherAsync>d__19.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at NuGet.PackageManagement.NuGetPackageManager.<PreviewInstallPackageAsync>d__43.MoveNext()
   --- End of inner exception stack trace ---
   at NuGet.PackageManagement.NuGetPackageManager.<PreviewInstallPackageAsync>d__43.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.PackageManagement.NuGetPackageManager.<InstallPackageAsync>d__28.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at NuGet.CommandLine.InstallCommand.<InstallPackage>d__34.MoveNext()<---
```

This what I have with 3.3:

```
Feeds used:
  http://ProGet/nuget/Default

Attempting to gather dependencies information for package 'A.1.1.1' with respect to project 'C:\Users\test\Downloads', targeting 'Any,Version=v0.0'
Attempting to resolve dependencies for package 'A.1.1.1' with DependencyBehavior 'Lowest'
Resolving actions to install package 'A.1.1.1'
Resolved actions to install package 'A.1.1.1'
Adding package 'A.1.1.1' to folder 'C:\Users\test\Downloads'
Added package 'A.1.1.1' to folder 'C:\Users\test\Downloads'
Successfully installed 'A 1.1.1' to C:\Users\test\Downloads
```

**Comment by NN---** [comment]
This is HTTP request of 3.4:

```
GET /nuget/Default/Packages(Id='A',Version='1.1.1') HTTP/1.1
```

And this is of 3.3

```
GET /nuget/Default/Packages(Id='A',Version='1.1.1.0') HTTP/1.1
```

ProGet differs 1.1.1 from 1.1.1.0 that's why it doesn't work with NuGet 3.4.

**Comment by yishaigalatzer** [comment]
That looks like a bug in ProGet, the ID is identical. I don't want to fix this in NuGet this late close to the release of 3.4, either fix the packages or contact ProGet to update their API

**Comment by NN---** [comment]
If it is bug in ProGet then It is a bug in all older NuGet versions.
This change makes a regression.
Supposing I had a package 1.1.1.0 , after installing it, it added using ".targets" file in my project with path packages\A-1.1.1.0.
With 3.4 I get a folder packages\A-1.1.1 and my code won't compile.

In case this will be a default behavior with no regrets , please document it as a breaking change since now you cannot install a package which you could.
Also it means that all tools (like CoApp) must be updated to always normalize version.

**Comment by yishaigalatzer** [comment]
> If it is bug in ProGet then It is a bug in all older NuGet versions.

The bug is that the server doesn't respect normalized versions, nuget is making an API call with a normalized version.

> With 3.4 I get a folder packages\A-1.1.1 and my code won't compile.

I'm not aware of such an issue, that is a breaking change and we will fix it in 3.3.

**Comment by NN---** [comment]
I see what you mean.
Tell me if I have package versions 1.1.1.0 should it be installed to A-1.1.1.0. folder or A-1.1.1 ?
Or perhaps it depends what version I specify to I install it?

**Comment by apxltd** [comment]
I just checked w/ the team here. 

We had to specifically implement the “4th part quirk” because NuGet (as of 3.3 and earlier treated) considers these as different versions. A significant number of packages and users relied on this behavior (and we received a lot of complaints about this) which is why we specifically added it.

Note that NuGet 3.3 and earlier also has the “leading zeros quirk” (for example, 1.01.0 ?= 1.1.0, depending on the query), but we chose not to implement that because only handful of packages (including, I think Ninject or Rhino or something?) used leading zeros to differentiate packages.

We agree in principle that "1.1.1 == 1.1.1.0" are identical. But since NuGet 3.3 and earlier (as well as lots of users) disagree, our principle didn't really matter here... however we did stick to our principle that "1.1.1 == 1.01.1". ProGet does not support that, where asNuGet does.

To be honest we really didn't want to have to implement the "4th part quirk", and would greatly prefer using semantic versioning (like we do with other package formats), but it's been there since NuGet 1.x, and people use and rely on it. I agree with @NN that this is a breaking change and should be documented as such. It's kind of a big one.

We've been getting quite a few complaints about this one already from 3.4-RC users, so it's gonna be a bit of a pain for them, because all we can tell them is "don't upgrade or don't rely on that behavior".

**Comment by yishaigalatzer** [comment]
Hey Alex, I guess the ask is to respect the query but return the metadata as is in the nuspec, that wouldnt be a breaking change.

Will double check on our side.

**Comment by apxltd** [comment]
I'm not sure I understand? NuGet 3.3 and earlier already supports A-1.1.0 and A-1.1.0.0 in the same feed. So, they are currently considered different packages.

I think the existing behavior is strange... but seems like it'd be even stranger to specify:
- 1.1.0 is returned if you ask for 1.1.0.0, unless there is already a 1.1.0.0 package
- 1.1.0.0 is returned if you ask for 1.1.0, unless there is already a 1.1.0 package
- PACKAGE_NOT_FOUND is returned if you ask for 1.0, even if there is 1.0.0

And then this will be even stranger when you consider "connectors" that aggregate multiple feeds. Which basically, everyone in a environment uses.

We really can't drop support for the "4th part quirk" in ProGet anytime soon, because many people are still on NuGet 2.x and rely on this behavior. At best, we could do a hack like "enable 4th part quirk mode" on a feed.

If you make this change, then it's a breaking API change. NuGet users are used to that. But please please please strongly document it in the release notes as a breaking change, so we can at least point the onslaught of users to that instead and advise them not to upgrade or not to rely on the behavior.

**Comment by yishaigalatzer** [comment]
That's not what we are saying:

1.1 is the same as 1.1.0 in terms of version equality, you cannot and should not have both of them on the same feed. Never

However when you restore a package nuget2 uses the nuspec ID to put it in the solution packages folder. Nuget 3 does the same, but normalizes it in the global packages folder.

There is no change here other than allowing a search/download request to find the package based on the normalized version. You still need to return the original ID from the nuspec.

If you still have concerns let's get on Skype/phone and hash because we are locking down the rtm release and id like to address this if I'm in the wrong

**Comment by yishaigalatzer** [comment]
> I think the existing behavior is strange... but seems like it'd be even stranger to specify:

This should be instead

Ask for 1, 1.0, 1.0.0, 1.0.0.0? Allow for a single package with that ID (because they are identical), return metadata with the id as it is specified in the nuspec

> 1.1.0 is returned if you ask for 1.1.0.0, unless there is already a 1.1.0.0 package

Return whatever is in the nuspec

> 1.1.0.0 is returned if you ask for 1.1.0, unless there is already a 1.1.0 package

Return whatever is in the nuspec

> PACKAGE_NOT_FOUND is returned if you ask for 1.0, even if there is 1.0.0

Package is returned with the id in the nuspec

> NuGet 3.3 and earlier already supports A-1.1.0 and A-1.1.0.0 in the same feed. 

That is not truly supported, and is not a scenario we should keep going, aggregators can fail or pick one of the packages. Similarly to what if two feeds have the same id + version with different content, that is not supported scenario

**Comment by NN---** [comment]
> 1.1 is the same as 1.1.0 in terms of version equality, you cannot and should not have both of them on the same feed. Never

Who has to enforce this ?
Currently if I run web server with NuGet.Server package (2.8.6 the latest) , I can easily put packages 1.1.1 and 1.1.1.0 in the folder.
NuGet 3.3 in VS 2015 shows me both packages with versions 1.1.1 for both.
The client doesn't enforce versions , it just relies on the server , which seems not good enough to me.

As well since the server should prevent it, does NuGet.Server has a bug too ?

**Comment by yishaigalatzer** [comment]
I need to check but I believe NuGet.Server only allows a single version @maartenba

**Comment by yishaigalatzer** [comment]
Tried to verify this with nuget.server 2.8.60717.93 I see the same problem. I think it is unfortunate and we will have to fix it in the client, as it is too wide spread.

However - I think the servers should be compliant with any version, because in project.json scenarios, the user can just type a normalized version AND we do normalize them when we use the global packages folders.

So there are two things to do here:
1. Make the servers follow the suggestion above and block duplicates of normalized versions @maartenba - please verify on the latest nuget.server (where this might be fixed because we should use the nuget.versioning package, and the gallery)
2. @zhili1208 we need to get this back into 3.4 RTM and get it fixed, it will be too big of a breaking change to handle.

**Comment by emgarten** [comment]
The client treats `1.1`, `1.1.0`, `1.01.0` and `1.1.0.0` as the same version using SemVer rules. The reason non-normalized versions were special cased in the past is because for v2 http calls the client would first send the version string exactly as the user specified it. If the package was not found the client would then start trying all other permutations to "help" the user out in the scenario where they typed the version wrong.

When the user installs a package without giving a version (to install the highest) the behavior is going to be random. The client will call `FindPackagesById` then order the packages using SemVer ordering and take FirstOrDefault which could be either one. It is the same today when installing from a local folder.

In 3.4.0-RC the client requests the normalized package version, if it doesn't exist it then falls back to FindPackagesById. This means two requests for every package that isn't on the server at all. With proget the behavior I was seeing with @zhili1208 is that when a `/Packages` request failed it would redirect to html which caused the feed parser to throw instead of falling back to FindPackagesById.

There are two ways the client can fix this:
1. Remove all calls to `/Packages` and call only `FindPackagesById` instead, this is how DNU handled things and NuGet does the same for project.json restore.
2. Add the NuGet.Core.dll behavior back where brute force calls are made to `/Packages` like before. This still doesn't work for packages with leading zeros and it means that for every not found package 3 requests will be made.

I suggest we go with the first option and deprecate `/Packages`. It will reduce the number of http calls and ensure that the client always finds the package. Scenarios where there are multiple non-normalized versions may hit issues, but that is going to happen either way.

As for NuGet.Server, it needs to be improved for these scenarios and is not a good example to go by. NuGet.Server is very minimal and is essentially just a local folder source served over http. It was only around 2.8.4 or so that a bug was fixed to allow NuGet.Server to actually order packages using SemVer. Before that versions were treated as strings and packages such as 1.0.9 would show up as the latest over 1.0.10.

**Comment by yishaigalatzer** [comment]
The nuget.server reference is that it doesn't understand normalized versions either and it should. Given that it doesn't ID go with @emgarten suggestion 1

**Comment by apxltd** [comment]
> 1.1 is the same as 1.1.0 in terms of version equality, you <s>cannot and</s> should not have both of them on the same feed. Never

I fully agree here with the "should not... never", but we had to implement it because a lot of users were relying on this versioning quirk when migrating from NuGet.Server or a share drive to ProGet, and it was a showstopper to not have this "feature" in their feed.

The main abuse of this quirk seems to be quality related: e.g. "1.0" is the first version, "1.0.0" is a patch", and so on. It's some nonsense like that. We've tried to educate folks _not_ to use that quirk, but they have a giant feed/infrastructure with dependencies everywhere. 

We did _not_ implement the leading zero quirk (1.01.1 ?= 1.1.1), as the majority of the usecases for that were working around the sorting quirk (1.10 < 1.9), which was since fixed. I _think_ the latest NuGet.Server still treats them differently, but I think it's ok to break support for that.

> Make the servers follow the suggestion above and block duplicates of normalized versions 

The main problem is existing packages; some of these orgs have like 100k's of packages. So even if they get blocked from being added, what to do about the existing packages? Which "A-1" is the real "A-1" (A-1.0, A-1.0.0, A-1.0.0.0)?

## 

Here's my quick thoughts: what about say, "SemVer Only" when using the v3 API, and keep the quirks in the v2 API. Or if not SemVer, then a similarly strict standard.

Seems like this will be the easiest to implement for you guys as well? So far as I can tell from looking at the v3 code a while ago, doesn't the v3 client just bundle a v2 client or something? That thing is so brittle, and I think _every single v2 release_ introduced different regressions (many can't upgrade past v2.6 NuGet.exe client still), so avoiding changing v2 will probably save everyone's sanity.

We haven't implemented a v3 API (nor have our friends at Artifactory or Sona), and I believe MyGet's support is still considered experimental. So, hardly anyone is using v3 for their in-house (maybe they forked NuGetGallery3, but that's really rare anymore), and anyone who relies on the quirk in v3 can just change their usage.

For ProGet users, we would have a nice transition then. Migrate to a v3 feed, and only semver packages are recognized.

**Comment by yishaigalatzer** [comment]
That just wouldn't work. We are throwing away the old v2 client as it is not sustainable. You are describing things predating even semver 1.

I guess these folks would have to stay on older versions, but we are not and should not support this going forward, regardless of the api is v1 or v2.

I'm OK with making the change to send the exact version the user requested, but we will not support anyone using this quirk on new version of 2015, by that I mean a single packages folder would not be guaranteed to work in this case and if it does that's "luck" and nothing intentional.

Also note that for any project using project.json that is broken by definition. The folder layout is version normalized and that's how we access packages.

The only deviation from semver is allowing the 4th number  in a version.

So in summary I think we want to declare this a breaking change even if slight for most folks and call it a day. Nuget needs to move off the crippling legacy modes from 4 years ago.

For the folks with 1000s of packages, I think they will have to write a script to fix them and their projects if they want to move along.

**Comment by NN---** [comment]
@yishaigalatzer It is not a problem to write a script and normalize all versions for future revisions, the question is whether you can pull packages with older revisions where they ask for non normalized versions successfully ?

**Comment by yishaigalatzer** [comment]
I'm suggesting we allow for that, but we don't allow for duplicates

**Comment by NN---** [comment]
@yishaigalatzer That is reasonable for me. I don't have duplicates.
Actually the version with 4th zero was an accident and wasn't intended at all.
But anyhow it helped to find a breaking change.

**Comment by apxltd** [comment]
> So in summary I think we want to declare this a breaking change even if slight for most folks and call it a day. Nuget needs to move off the crippling legacy modes from 4 years ago.

:+1: -- we'll see how many folks notice, and enforce it on the ProGet end at some point (maybe as an option, maybe not). I think it's easy enough simply not to upgrade if it's that big of a deal.

**Comment by NN---** [comment]
@apapadimoulis Will it be possible given package 1.1.0.0 to ask for 1.1 and get the package successfully ?

**Comment by apxltd** [comment]
@NN--- I'm afraid not, because then that would be a breaking change on the server side, as 3.4 and pre-3.4 resolution logic is incompatible. 

So for now, I would suggest to simply not upgrade to 3.4 if you rely on pre-3.4 logic.

**Comment by NN---** [comment]
It is not a problem for me to add the same package of 1.1.1.0 with version 1.1.1 and have both in the server.
That way both pre-3.4 and 3.4 will work.

I don't know how many people have **really different** packages with trailing zeros.
My guess that most just didn't bother with normalizing like me since it just worked.

**Comment by yishaigalatzer** [comment]
@apapadimoulis that is not a major breaking change, and you should absolutely support it. Duplicate packages of the same version is a non-scenario.

**Comment by apxltd** [comment]
@yishaigalatzer I'd thought so too... except there were a lot of folks that had A-1, A1.0, A-1.0.0, etc. in their feed, so we had to implement it to allow them to migrate or use connectors to existing feeds...

But I agree, they probably aren't relying on "1.0.0" returning PACKAGE_NOT_FOUND if there is a "1.0.0.0" in the feed. So, we could ship that as a 3.4-compatiability maintenance version if needed.

So just to make sure I understand the specs...
- The 3.4 client will always send a "normalized version" in the query.
- A "normalized version" has 4 parts (0.0.0.0)
- If a non-normalized version" has less than 4 parts, you can normalize it by appending .0's to it

**Comment by yishaigalatzer** [comment]
Short term, we plan to improve the 3.4 behavior as follows - And we will put bits in your hands as soon as they are ready (on the myget nugetbuild feed).

3.4 client will send a normalized version first which is 3 digits or 4 as applicable. 
For example:

| User input | Id in nuspec | Request | Response |
| --- | :-: | --: | --- |
| 1.0 | 1.0.0.0 | 1.0.0 | Value in nuspec |
| 1.00 | 1.0 | 1.0.0 | Value in nuspec |
| 1.00.0.1 | 1.0.0.1 | 1.0.0.1 | Value in nuspec |

Also we identified that the root issue in the proget repo is that it returns a 200/html instead of 404 for that request. So lets get that fixed as well.

If we get a 404 we will get all versions of the packages and pick it from there (this is already in the code but failed because of the 200/html response)

**Comment by joelverhagen** [comment]
> @zhili1208 we need to get this back into 3.4 RTM and get it fixed, it will be too big of a breaking change to handle.

We are now observing HTTP status codes correctly for V2 feeds with https://github.com/NuGet/NuGet.Client/pull/374.

**Comment by maartenba** [comment]
NuGet.Server now also supports normalized versions.

**Comment by joelverhagen** [comment]
I applied the fix, as mentioned above and @zhili1208 validated this on NuGet.org, ProGet, and NuGet.Server.

**Comment by fahadash** [comment]
Is this fix up on the VS Extensions and Updates manager? I looked for updates on Nuget Package Manager and found nothing.

**Comment by yishaigalatzer** [comment]
@fahadash  Yes, what version do you have installed?

**Comment by fahadash** [comment]
@yishaigalatzer  I have 3.4.4.1321 . I installed VS 2015 Update 2 and it works fine now. Thanks.
